src/platform_impl/linux/wayland/window/state.rs

```diff
@@ -725,17 +728,32 @@ impl WindowState {
     }
 
     fn apply_custom_cursor(&self, cursor: &CustomCursor) {
-        self.apply_on_pointer(|pointer, _| {
+        self.apply_on_pointer(|pointer, data| {
             let surface = pointer.surface();
 
-            let scale = surface.data::<SurfaceData>().unwrap().surface_data().scale_factor();
+            let scale = if let Some(viewport) = data.viewport() {
+                let scale = self.scale_factor();
+                viewport.set_destination(
+                    (cursor.w as f64 / scale).round() as i32,
+                    (cursor.h as f64 / scale).round() as i32,
```

Comment:
Could you use the same formula we use elsewhere for fractional scaling? You can follow it from `inner_size` on the window.